web/lib/supabase/utils.ts

```diff
@@ -0,0 +1,256 @@
+import {
+  SupabaseClient,
+  TableName,
+  RowFor,
+  ViewName,
+  DataFor,
+} from 'common/supabase/utils'
+import { useEffect, useState } from 'react'
+import { uniqBy } from 'lodash'
+import { filterDefined } from 'common/util/array'
+import { RealtimeChannel } from '@supabase/realtime-js'
+
+function hasFsUpdatedTime(obj: any): obj is { fs_updated_time: number } {
+  return 'fs_updated_time' in obj
+}
+
+let mainChannel: undefined | RealtimeChannel = undefined
+let subscriptionStatus:
+  | 'SUBSCRIBED'
+  | 'TIMED_OUT'
+  | 'CLOSED'
+  | 'CHANNEL_ERROR'
+  | 'CONNECTING' = 'CONNECTING'
+const callbacks: {
+  hookId: string
+  callback: (data: any) => void
+  filter: string
+  table: string
+}[] = []
+let mainChannelResubscribeInterval: NodeJS.Timer | undefined = undefined
+
+export function useValuesFromSupabase<
+  T extends TableName | ViewName,
+  K extends keyof RowFor<T>
+>(
+  table: T,
+  rowGroupKey: K,
+  rowGroupValue: string,
+  uniqueRowDataKey: K,
```

Comment:
This should always be the same for any table name, right? You could make a map from table name to primary keys, so that people don't need to pass it into this code.